File: static/shared/ts/components/workspace-viewer/_security.ts
/** Keep an untrusted value on one physical log line. */
export function sanitizeLogValue(value: string): string {
  return value
    .replace(/[\r\n\u2028\u2029]/g, " ")
    .replace(/[\u0000-\u001f\u007f]/g, "?");
}

File: apps/security.py
"""Small security helpers shared by HTTP-facing applications."""


def safe_log_field(value: object, *, limit: int = 512) -> str:
    """Render an untrusted value as one bounded physical log line."""
    return str(value).replace("\r", "\\r").replace("\n", "\\n")[:limit]
